engineering/standards/frontend.md

```diff
@@ -151,11 +151,15 @@ Note: some of our older code still does not conform to this standard.
 
 ### Component spacing and layout
 
-Components should not set any space around their outer-most visual edges — instead the consuming component should be responsible for layout and spacing.
+To preserve modularity and encapsulation, components should not define their own layout-related behavior — instead the consuming component should be responsible for layout and spacing. Components should not set any space around their outer-most visual edges or define their own z-index.
 
 - **margin**: The component's root element should not set any margin.
-
 - **padding**: If the component's root element has border, it's fine to set padding because the border will serve as the outer-most visual edge. But if there's no border, then there should be no padding.
+- **z-index**:
+    - The component's root element should not set any z-index.
+    - It's fine for child elements _within the component_ to set a z-index, but in such cases the component's root element must establish its own [stacking context](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context). It's best to use `isolation: isolate;` to establish the stacking context because it clearly communicates intent and works without setting z-index. Once your component has its own stacking context, then you're free to set the z-index values within the component without thinking about anything outside the component. You can use simple values like `1` and `2` within the component because everything is encapsulated.
```

Comment:
We may want to establish some css variables with default z-index values for use with components within the component library. I think it's reasonable for components like modal and dropdown content to have z-index values.

Reply:
Let me see if I follow you...

1. The Modal system would set:

    ```css
    .modal {
      z-index: var(--modal-z-index, auto);
    }
    ```

1. The Toast system would set:

    ```css
    ..toast-presenter {
      z-index: var(--toast-presenter-z-index, auto);
    }
    ```

1. In `src/component-library/common/styles/variables.scss`, we'd set something like:

    ```css
    :root {
      --modal-z-index: 100;
      --toast-presenter-z-index: 101;
    }
    ```

    This way, toast messages appear above the modal overlay.

Is that what you had in mind? If so, I think that's good. I would just want to confirm that any CSS we define within `App.css` will have higher precedence than that defined within that `variables.scss` file -- because component library consumers will need a mechanism to override those z-index values.

We'll need to play with Dropdown a bit to make sure everything works -- i.e. a dropdowns opened outside of modals should display under modals and and dropdowns opened within modals should display above modals. My intuition tells me that modals and dropdowns will need to have the same z-index value. I'm not positive that'll cover all the use-cases though.

Reply:
> Is that what you had in mind

Yes, this is what I had in mind.